designs/2024-ignore-file-directive/README.md

```diff
@@ -0,0 +1,121 @@
+
+
+Many libraries that generate code generate code with leading `/* eslint-disable */` directives at the top of generated files. However, those files are still parsed and validated, which slows down the linting process needlessly. Potentially large and complex files are processed, even though their results are useless to the application developer. In the worst case, bugs in linting rules (see [this issue about a broken linting rule](https://github.com/eslint/eslint/issues/18022)) crash the entire ESLint process over artifacts an application developer has no control over. The problem is that `/* eslint-disable */` merely mutes linting results.
+
+This directive will work alongside the `.eslintignore` file in that it too will indicate which file be parsed or not. It extends the concept of ignoring source files by allowing files (specifically generated files) to exclude themselves.
+
+Lastly, library devlopers seldom actually mean "parse, but mute errors" when `/* eslint-disable */` is added at the top of files. While it the best option currently available, "completely skip this file" is actually meant.
+
+To summarize:
+
+1. Performance boost by not parsing unwanted files
+2. Don't crash over code the application developer has no/limited control over
+3. Does what leading `/* eslint-disable */` are meant to do
+
+## Detailed Design
+
+Please note that I am not familiar with the internals of ESLint and might make assumptions that are incorrect.
+
+For any _source file_ (the root of an AST), if it contains a leading `/* eslint-ignore-file */` directive, skip parsing it. More concretely, the following pseudocode algorithm explains how to determine whether a file should be ignored or processed:
+
+```
+for line of source file:
+  if line (is not zero or more whitespace) and (does not start with zero or more whitespace + "/"):
+    parse file
+  if line is "/* eslint-ignore-file */":
+    skip file parsing
+```
+
+A more sophisticated algorithm may be chosen at a later point in the lifecycle of this RFC.
+
+In practice, the imaginary
+
+```ts
+export declare function parseFile(filePath: string): AST;
+```
+
+needs to be replaced with:
+
+```ts
+export function parseFile(filePath: string): AST | null {
+  const fileContent = streamFileContent(filePath);
+
+  if (shouldIgnore(fileContent)) return null;
+  return parseFileImpl(fileContent.collectToString());
+}
+
+declare function parseFileImpl(fileContent: string): AST;
+```
+
+How the rest of the system deals with `null` results from parsing a source file is not in the scope of this RFC.
+
+If an `/* eslint-ignore-file */` directive is encountered while parsing a file, a non-fatal error shall be raised, and the directive shall have no further effect.
```

Comment:
I thought the point was that we wouldn't parse the file? But here you're saying we will in order to find the comment? Can you clarify?

And what does it mean for a non-fatal error to be raised? Do you mean it would be look like a lint violation? Would the user get any output?